pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+provide TLS functionality that is not so strongly tied to OpenSSL. It also
+proposes to update standard library modules to use only the interface exposed
+by these abstract base classes wherever possible. There are three goals here:
+
+1. To provide a common API surface for both core and third-party developers to
+   target their TLS implementations to. This allows TLS developers to provide
+   interfaces that can be used by most Python code, and allows network
+   developers to have an interface that they can target that will work with a
+   wide range of TLS implementations.
+2. To provide an API that has few or no OpenSSL-specific concepts leak through.
+   The ``ssl`` module today has a number of warts caused by leaking OpenSSL
+   concepts through to the API: the new ABCs would remove those specific
+   concepts.
+3. To provide a path for the core development team to make OpenSSL one of many
+   possible TLS backends, rather than requiring that it be present on a system
+   in order for Python to have TLS support.
+
+The proposed interface is laid out below.
+
+
+Abstract Base Classes
+---------------------
+
+There are several interfaces that require standardisation. Those interfaces
+are:
+
+1. Configuring TLS, currently implemented by the `SSLContext`_ class in the
+   ``ssl`` module.
+2. Wrapping a socket object, currently implemented by the `SSLSocket`_ class
+   in the ``ssl`` module.
+3. Providing an in-memory buffer for doing in-memory encryption or decryption
+   with no actual I/O (necessary for asynchronous I/O models), currently
+   implemented by the `SSLObject`_ class in the ``ssl`` module.
+4. Specifying TLS cipher suites. There is currently no code for doing this in
+   the standard library: instead, the standard library uses OpenSSL cipher
+   suite strings.
+5. Specifying application-layer protocols that can be negotiated during the
+   TLS handshake.
+6. Specifying TLS versions.
+7. Reporting errors to the caller, currently implemented by the `SSLError`_
+   class in the ``ssl`` module.
+
+While it is technically possible to define (2) in terms of (3), for the sake of
+simplicity it is easier to define these as two separate ABCs. Implementations
+are of course free to implement the concrete subclasses however they see fit.
+
+Obviously, (4) doesn't require an abstract base class: instead, it requires a
+richer API for configuring supported cipher suites that can be easily updated
+with supported cipher suites for different implementations.
+
+Context
+~~~~~~~
+
+The ``Context`` abstract base class defines an object that allows configuration
+of TLS. It can be thought of as a factory for ``TLSWrappedSocket`` and
+``TLSWrappedBuffer`` objects.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+    ServerNameCallback = Callable[[TLSBufferObject, Optional[str], Context], Any]
+
+    class _BaseContext(metaclass=ABCMeta):
```

Comment:
Do we want to force that into the API? Or do we want to say that concrete implementations should wrap those details into their ClientContext or ServerContext objects?